### ISSUE TITLE
`no-useless-undefined`: false negative in TS file

### ISSUE THREAD

**Issue by ocavue** [sindresorhus/eslint-plugin-unicorn#3330]
Given the following TS code:

```ts
export function fn1(): undefined {
	return undefined;
}

export function fn2(): void {
	return undefined;
}

export function fn3() {
	return undefined;
}

export function fn4(): number | undefined {
  if (1+1 > Math.random()) {
    return 2;
  } else {
    return undefined;
  }
}

export function fn5(): number | void {
  if (1+1 > Math.random()) {
    return 2;
  } else {
    return undefined;
  }
}

export function fn6() {
  if (1+1 > Math.random()) {
    return 2;
  } else {
    return undefined;
  }
}

```

And enable `unicorn/no-useless-undefined` rule.

Current behavior: `fn1` `fn3` and `fn6` are reported as errors.
Expected behavioor: all 6 functions are reported as errors.

Minimal reproduction: https://github.com/issueset/no-useless-undefined-ts-issue.git

**Comment by vetrovk** [comment]
I investigated this on latest main.

The straightforward part is that : void is currently missed because isUndefinedReturnType() only checks TSUndefinedKeyword.

However, the union cases would change existing behavior: the current TypeScript tests intentionally allow return undefined for string | undefined, referencing #880.

Before opening a PR, could you confirm the intended behavior for union return types?

Should return undefined be reported for all top-level unions containing undefined or void, such as number | undefined, string | undefined, and number | void, or should this issue only cover explicit : void?
